pkg/skaffold/docker/parse.go

```diff
@@ -158,50 +159,73 @@ func processBaseImage(value *parser.Node) ([]string, error) {
 		logrus.Debugf("SCRATCH base image found, skipping check: %s", base)
 		return nil, nil
 	}
-	cfg, err := RetrieveConfig(base)
+	img, err := RetrieveImage(base)
 	if err != nil {
 		return nil, err
 	}
-	logrus.Debugf("Found onbuild triggers %v in image %s", cfg.Config.OnBuild, base)
-	return cfg.Config.OnBuild, nil
+	logrus.Debugf("Found onbuild triggers %v in image %s", img.Config.OnBuild, base)
+	return img.Config.OnBuild, nil
 }
 
-var imageToConfigCache sync.Map
+var imageCache sync.Map
 
-func retrieveImageConfig(image string) (*manifest.Schema2Image, error) {
-	cachedCfg, present := imageToConfigCache.Load(image)
+func retrieveImage(image string) (*manifest.Schema2Image, error) {
+	cachedCfg, present := imageCache.Load(image)
 	if present {
 		return cachedCfg.(*manifest.Schema2Image), nil
 	}
 
-	ref, err := docker.ParseReference("//" + image)
+	raw, err := retrieveLocalImage(image)
+	if err != nil {
+		raw, err = retrieveRemoteImage(image)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	cfg := &manifest.Schema2Image{}
+	if err := json.Unmarshal(raw, cfg); err != nil {
+		return nil, err
+	}
+
+	imageCache.Store(image, cfg)
+
+	return cfg, nil
+}
+
+func retrieveLocalImage(image string) ([]byte, error) {
+	// TODO(dgageot): docker api might not be available or it could be minikube
+	api, err := NewImageAPIClient()
```

Comment:
This is tricky since we've made an implicit dependency between getting the docker daemon and the cluster context.

In the future, we should really do this clientside and not talk to the daemon, getting around this issue. But for now we should defer to the daemon for pull.

It makes sense to have a function that does the resolution, so other packages can call it. Essentially moving the switch statement outside local.go into its own function.

```golang
func NewImageAPIClient() (ImageAPIClient, error) {
// get kubectl context
// context switch statement
}
```

This makes the package dependency more explicit and has the question of where it should live. Its maybe not totally correct to have pkg/docker depend on pkg/kubernetes, but I don't think we have a choice by what we've chosen.

Alternatively, we could use a daemonless pull library like https://github.com/containers/image or https://github.com/google/go-containerregistry. But I think that can wait until we run into an import issue.

Reply:
That function could be memoized too, so that contexts dont change midway through a skaffold invocation

Reply:
Thanks @r2d4, I'll try something based on your idea. To be precise, we already use a daemonless pull library. It's just that it's both slow and incorrect compared to reading from the daemon's local storage.